### PR TITLE
feat: subscription to use read type tx

### PR DESCRIPTION
**Pull request by Shubham8287**
# subscription actor to use read tx.
1. In datastore, made assigned `traits::Tx::Tx` to point to`TxId` rather `MutTxId`.
2. Done required changes in `relational_db.rs` to support 1.
3. started using`Tx` in subscription actor.
4. refactored `vm.rs` to work with both types of tx.
5. refactored `execute.rs`

broken test will be fixed by https://github.com/clockworklabs/SpacetimeDB/pull/695

### CODE REVIEW

File: crates/core/src/db/relational_db.rs
Comment (cloutiertyler) [inline crates/core/src/db/relational_db.rs:279]: 👍

File: crates/core/src/db/relational_db.rs
Comment (cloutiertyler) [inline crates/core/src/db/relational_db.rs:288]:
```suggestion
        log::trace!("ROLLBACK MUT TX");
```

Just for consistency

File: crates/core/src/db/relational_db.rs
Comment (cloutiertyler) [inline crates/core/src/db/relational_db.rs:275]:
```suggestion
        log::trace!("BEGIN MUT TX");
```

Just for consistency

File: crates/core/src/db/relational_db.rs
Comment (cloutiertyler) [inline crates/core/src/db/relational_db.rs:205]: I suppose we could have this take a `TxMode` or something of a similar concept and not duplicate all the functions, but I dunno how I feel about that. I think what you have now is probably better, but thought I'd mention it.

File: crates/core/src/sql/execute.rs
Comment (cloutiertyler) [inline crates/core/src/sql/execute.rs:85]: Probably want to check with Mario or Joshua on this one if you haven't already, simply because I'm just not familiar.

File: crates/core/src/vm.rs
Comment (cloutiertyler) [inline crates/core/src/vm.rs:382]: Is this an error or is this actually unreachable? Like is it possible to have a TxMode::Tx if you've arrived at this function?

File: crates/standalone/src/lib.rs
Comment (cloutiertyler) [inline crates/standalone/src/lib.rs:586]:
```suggestion
```

You are correct, this should be a release here not a commit.

**Review by cloutiertyler — COMMENTED** [review]
Left some small comments but overall this is looking really really good. I would be okay to merge this once the one it's dependent on merges.

File: crates/core/src/db/relational_db.rs
Comment (Shubham8287) [reply crates/core/src/db/relational_db.rs:205]: I thought about that as well but then all callers have to keep wrapping tx objects in`TxMode` to pass as param, Right now, `TxMode` is local to `vm.rs` only.
I felt bit of wrapper methods duplicacy is fine here to avoid using extra enum everywhere.

File: crates/core/src/sql/execute.rs
Comment (Shubham8287) [reply crates/core/src/sql/execute.rs:85]: Let me add them as reviewer, specially on this snippet @joshua-spacetime , @mamcx

File: crates/core/src/vm.rs
Comment (Shubham8287) [reply crates/core/src/vm.rs:382]: This should be unreachable code.

File: crates/core/src/sql/execute.rs
Comment (mamcx) [reply crates/core/src/sql/execute.rs:85]: Yes this looks correct. `CrudExpr::Query(_)` only takes `SELECT..` expressions without mutate effects.

I suggest to put this logic in `CrudExpres` (similar to `CrudExpr::is_reads()`) instead of here. The idea is that here is an upper layer and all decisions should be inside the `vm` or `query` layer so is more visible to the people that do optimizations.

File: crates/core/src/vm.rs
Comment (mamcx) [inline crates/core/src/vm.rs:368]: If here is this then the upper layer should prove this is not possible. If can't it should be a normal `Err` instead (is better to avoid a crash)

File: crates/core/src/sql/execute.rs
Comment (Shubham8287) [reply crates/core/src/sql/execute.rs:85]: Agree, that makes sense

File: crates/core/src/vm.rs
Comment (Shubham8287) [reply crates/core/src/vm.rs:368]: Reaching to this branch anytime implies a bug in code and not to do anything with Users interaction.
I added a comment on `VmProgram::eval_query()` (only caller) to specify safetu gurantee.

File: crates/core/src/db/relational_db.rs
Comment (joshua-spacetime) [inline crates/core/src/db/relational_db.rs:536]: There's no need to add this metric since it doesn't have the desired behavior.

File: crates/core/src/sql/execute.rs
Comment (joshua-spacetime) [inline crates/core/src/sql/execute.rs:109]: Are compilation and execution in two separate transactions now?

File: crates/core/src/subscription/module_subscription_actor.rs
Comment (joshua-spacetime) [inline crates/core/src/subscription/module_subscription_actor.rs:177]: What does it mean to rollback a read transaction?

File: crates/core/src/sql/execute.rs
Comment (Shubham8287) [reply crates/core/src/sql/execute.rs:109]: yes, compilation works with read-only tx, then execution could be read-tx or write-tx based on compilation result (ast).

File: crates/core/src/subscription/module_subscription_actor.rs
Comment (Shubham8287) [reply crates/core/src/subscription/module_subscription_actor.rs:177]: I guess naming could be better here, `finish_tx` looks better. It's basically to drop the guard.

File: crates/core/src/sql/execute.rs
Comment (joshua-spacetime) [reply crates/core/src/sql/execute.rs:109]: I think that's fine. Can you just add a comment to that effect since it might be unexpected?

File: crates/core/src/subscription/module_subscription_actor.rs
Comment (joshua-spacetime) [reply crates/core/src/subscription/module_subscription_actor.rs:177]: Yeah I agree, `finish` instead of `rollback` is clearer.

**Review by joshua-spacetime — APPROVED** [review]
LGTM pending some minor nits and tests passing.

**Pull request merged**
